### ISSUE TITLE
Streamification

### ISSUE THREAD

**Issue by pimterry** [murat-dogan/node-datachannel#81]
Hi,

I'm starting to build on this library, and I'm finding that integrating data channels into my existing infrastructure is a little tricky. In large part, this is because they're conceptually like streams, but they're not actually node.js streams, so they don't neatly match up with any other APIs.

Would you be open to changing this? If a DataChannel was a [Duplex](https://nodejs.org/api/stream.html#implementing-a-duplex-stream) stream, then my code could easily pipe data between DataChannels and elsewhere, and I could write code that interchangeably worked on DataChannels or TCP sockets or unix pipes.

I've also hit issues with wanting to attach multiple listeners, and with messages being sent before `channel.onMessage` is attached and being lost. The standard stream behaviour (where there are no listener limits, and data is buffered if there are 0 listeners at any time) would fix this nicely.

Would you be open to a PR to change this behaviour? I think it should be fairly easy: it requires defining a Duplex that reads via onMessage and write via sendMessage/sendBinaryMessage, and I think I have most of an implementation of that already. The same probably applies to media tracks, but I haven't investigate those much yet.

I think it would be possible to do this entirely backward compatibly, either by converting the DataChannel itself into a stream or adding a `.stream` property, and have an implementation for that in pure JS, my only issue is I'm not sure how to make that work within the C++ implementation here.

**Comment by murat-sx** [comment]
Hi,
 
> Would you be open to changing this? If a DataChannel was a [Duplex](https://nodejs.org/api/stream.html#implementing-a-duplex-stream) stream, then my code could easily pipe data between DataChannels and elsewhere, and I could write code that interchangeably worked on DataChannels or TCP sockets or unix pipes.

There could be a way to add a stream property as an extra function, without infecting current mechanisms. Actually this could be done as an extra feature in an extra class/function. Creating an extra function (or functions) in `lib/index.js` should do the trick. Just creating a stream mechanism and embedding `DataChannel` class should be enough. But I am not sure if this would be a feature that somebody else would use. Implementing this mechanism in your code should be easy too.

> I've also hit issues with wanting to attach multiple listeners, and with messages being sent before `channel.onMessage` is attached and being lost. The standard stream behaviour (where there are no listener limits, and data is buffered if there are 0 listeners at any time) would fix this nicely.

Buffering the data is not a job for  `DataChannel`s actually. If you need this behaviour then I guess you will need an extra layer in your code.
Currently you can assign only one listener per `DataChannel` which is also standard behaviour for `libdatachannel`. For multiple listener we will need an extra layer I guess.

> Would you be open to a PR to change this behaviour? I think it should be fairly easy: it requires defining a Duplex that reads via onMessage and write via sendMessage/sendBinaryMessage, and I think I have most of an implementation of that already. The same probably applies to media tracks, but I haven't investigate those much yet.

For me, writing a seperate wrapper JS class in your code or in to the   `lib` folder is more logical.   

> I think it would be possible to do this entirely backward compatibly, either by converting the DataChannel itself into a stream or adding a `.stream` property, and have an implementation for that in pure JS, my only issue is I'm not sure how to make that work within the C++ implementation here.

I don't expect any changes to C++ code for this functionality

**Comment by murat-dogan** [comment]
Hi,
 
> Would you be open to changing this? If a DataChannel was a [Duplex](https://nodejs.org/api/stream.html#implementing-a-duplex-stream) stream, then my code could easily pipe data between DataChannels and elsewhere, and I could write code that interchangeably worked on DataChannels or TCP sockets or unix pipes.

There could be a way to add a stream property as an extra function, without infecting current mechanisms. Actually this could be done as an extra feature in an extra class/function. Creating an extra function (or functions) in `lib/index.js` should do the trick. Just creating a stream mechanism and embedding `DataChannel` class should be enough. But I am not sure if this would be a feature that somebody else would use. Implementing this mechanism in your code should be easy too.

> I've also hit issues with wanting to attach multiple listeners, and with messages being sent before `channel.onMessage` is attached and being lost. The standard stream behaviour (where there are no listener limits, and data is buffered if there are 0 listeners at any time) would fix this nicely.

Buffering the data is not a job for  `DataChannel`s actually. If you need this behaviour then I guess you will need an extra layer in your code.
Currently you can assign only one listener per `DataChannel` which is also standard behaviour for `libdatachannel`. For multiple listener we will need an extra layer I guess.

> Would you be open to a PR to change this behaviour? I think it should be fairly easy: it requires defining a Duplex that reads via onMessage and write via sendMessage/sendBinaryMessage, and I think I have most of an implementation of that already. The same probably applies to media tracks, but I haven't investigate those much yet.

For me, writing a seperate wrapper JS class in your code or in to the   `lib` folder is more logical.   

> I think it would be possible to do this entirely backward compatibly, either by converting the DataChannel itself into a stream or adding a `.stream` property, and have an implementation for that in pure JS, my only issue is I'm not sure how to make that work within the C++ implementation here.

I don't expect any changes to C++ code for this functionality

**Comment by paullouisageneau** [comment]
@murat-dogan I'm not sure adding an extra layer is the simplest way here. For instance, regarding buffering, it is actually done by libdatachannel: when the onMessage callback is not set, messages are buffered. This behavior is lost with node-datachannel as the onMessage callback is never removed internally, leading the messages being discarded instead. In that case, I think it would be way easier to modify the logic a bit rather than re-implementing it in an extra layer.

**Comment by murat-dogan** [comment]
Regarding the buffering, I didn't notice that and yes this logic could be changed. I will check it. 
For multiple listener and stream behavior I think extra layer will be easier.

**Comment by pimterry** [comment]
Thanks to you both for your replies, all excellent points.

I don't think it's useful to focus too much on the buffering here though. That would be a helpful feature, but it's a symptom of the general problem really: a DataChannel is clearly a readable & writable stream of data,, and most idiomatic node code represents things like that as duplex streams (TCP sockets, compression, crypto, file IO, TLS connections, HTTP requests & responses, HTTP/2 sessions, WS's websockets, etc etc etc) but this library doesn't.

Providing a stream interface here would provide all the build-in features of node's streams, basically for free: buffering in both directions (with configurable buffer limits, so this remains optional), plus backpressure to automatically avoid overloading downstream streams, easy pipelining to process input & output to data channels, and lots of useful events that listeners can attach to in a reliable standard way.

It would also mean that they'd integrate easily into those other types of streams (pipe data from a WebRTC data channel into another data channel whilst also logging it to a file, in one line of code!) and would let you use the thousands of other stream-handling libraries to do things like filter streaming data, merging multiple streams, define pipelines of data transformations, and coordinating error handling across sets of streams.

I think you're right @murat-dogan and for now at least it makes sense for this to be optional, but I think it would be a really useful option to have here that would be useful to others, because node streams are very very widely used (as one data point, [merge-stream](https://www.npmjs.com/package/merge-stream) has 22 million downloads a week).

The implementation I have in my own code so far is very short (<100 lines with comments and everything) and there's zero dependencies - how about I finish testing it and then open a PR to add it as an optional feature in index.js, and we go from there?

**Comment by murat-dogan** [comment]
@pimterry For me it is completely OK. To write the stream implementation in a separate file and export it from `index.js` could be better if you don't mind.

**Comment by paullouisageneau** [comment]
> After v0.0.24 the behavior has changed for all callbacks. If user does not call `onMessage` callback then `node-datachannel` will not call it. If `libdatachannel` does buffering then it should work.

Indeed, it should work fine in the beginning when no callback is set, my point was more precisely that unless I'm mistaken there is no way to unset the callback from JavaScript (for instance when the number of listeners reaches zero). However, buffering in JS could be relatively straightforward too.

Anyway, the feature looks great!
